### PR TITLE
IGNITE-23707 Optimize update() in hybrid clocks

### PR DESCRIPTION
**Pull request by vldpyatkov**
```
Before
Benchmark                                            Mode  Cnt       Score      Error  Units
HybridClockBenchmark.hybridClockNowFiveThreads      thrpt   10   37536.317 ± 2033.320  ops/s
HybridClockBenchmark.hybridClockNowSingleThread     thrpt   10  238352.974 ± 4753.761  ops/s
HybridClockBenchmark.hybridClockNowTenThreads       thrpt   10   34136.852 ±  283.247  ops/s
HybridClockBenchmark.hybridClockUpdateFiveThreads   thrpt   10   14143.705 ±  365.301  ops/s
HybridClockBenchmark.hybridClockUpdateSingleThread  thrpt   10  124273.226 ± 3827.199  ops/s
HybridClockBenchmark.hybridClockUpdateTenThreads    thrpt   10    9028.719 ±  249.099  ops/s


After
Benchmark                                            Mode  Cnt       Score      Error  Units
HybridClockBenchmark.hybridClockNowFiveThreads      thrpt   10   35338.271 ± 5234.633  ops/s
HybridClockBenchmark.hybridClockNowSingleThread     thrpt   10  253800.841 ± 1555.133  ops/s
HybridClockBenchmark.hybridClockNowTenThreads       thrpt   10   35570.363 ±  581.248  ops/s
HybridClockBenchmark.hybridClockUpdateFiveThreads   thrpt   10   56725.796 ± 1055.919  ops/s
HybridClockBenchmark.hybridClockUpdateSingleThread  thrpt   10  241928.824 ± 3441.524  ops/s
HybridClockBenchmark.hybridClockUpdateTenThreads    thrpt   10   53343.884 ±  136.487  ops/s
```

https://issues.apache.org/jira/browse/IGNITE-23707

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java [hunk 1]
```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
...
+            if (oldLatestTime >= requestTimeLong) {
+                return hybridTimestamp(LATEST_TIME.incrementAndGet(this));
+            }
 
-            // TODO https://issues.apache.org/jira/browse/IGNITE-23707 avoid CAS on logical part update.
+            long now = currentTime();
+
+            if (now > requestTimeLong) {
+                if (LATEST_TIME.compareAndSet(this, oldLatestTime, now)) {
+                    return hybridTimestamp(now);
```
Comment (rpuch) [inline modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:128]: Why don't we notify listeners in these cases? Clock is still moved forward

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java [hunk 2]
```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
...
+                }
+            } else {
+                long newLatestTime = requestTimeLong + 1;
 
-            if (LATEST_TIME.compareAndSet(this, oldLatestTime, newLatestTime)) {
-                notifyUpdateListeners(newLatestTime);
+                if (LATEST_TIME.compareAndSet(this, oldLatestTime, newLatestTime)) {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:133]: Jira says that we should do an atomic increment in this branch:
> We can skip CAS if only logical part of clocks need to be updated.
This can be achieved using atomic increment, which is more efficient.

I still see a CAS here. Please replace it with `incrementAndGet`, or explain why can't do this.

Do we have any benchmark results here?

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java [hunk 3]
```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
...
 
+        while (true) {
             // Read the latest time after accessing UTC time to reduce contention.
-            long oldLatestTime = latestTime;
+            long oldLatestTime = this.latestTime;
 
-            long newLatestTime = max(requestTime.longValue() + 1, max(now, oldLatestTime + 1));
+            if (oldLatestTime >= requestTimeLong) {
+                return hybridTimestamp(LATEST_TIME.incrementAndGet(this));
+            }
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:122]: I don't see this change being mentioned in Jira. Are we sure it's safe? Maybe it needs a comment

**Review by rpuch — COMMENTED** [review]
> We can skip CAS if only logical part of clocks need to be updated.
> This can be achieved using atomic increment, which is more efficient.

Why is it more efficient? Doesn't such an update use CAS under the hood? According to the code of `AtomicLongFieldUpdater`, it does:

```
    public long incrementAndGet(T obj) {
        long prev, next;
        do {
            prev = get(obj);
            next = prev + 1;
        } while (!compareAndSet(obj, prev, next));
        return next;
    }
```
Is there some kind of an intrinsic?

Also, does this change bring any improvement? Could you please include the figures in a PR comment (excluding the lock-related change)?

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java [hunk 4]
```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
      */
     @Override
     public final HybridTimestamp update(HybridTimestamp requestTime) {
-        while (true) {
-            long now = currentTime();
+        long requestTimeLong = requestTime.longValue();
 
+        while (true) {
             // Read the latest time after accessing UTC time to reduce contention.
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:117]: Yes, I moved it below because current time is not required in some cases where we break the loop.
I will remove this comment.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 1 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:128]: Yes, but here the clock is not changed by jumping up. It means it is not faster than the scheduler in clock waiter.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 3 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:122]: It seems it is the obvious requirement. The clock current time has to be bigger than the value in parameters. The minimum change is incrementing the old value in the clock.

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/HeapLockManager.java [hunk 5]
```diff
@@ -174,7 +174,7 @@ public CompletableFuture<Lock> acquire(UUID txId, LockKey lockKey, LockMode lock
         if (lockKey.contextId() == null) { // Treat this lock as a hierarchy(coarse) lock.
             CoarseLockState state = coarseMap.computeIfAbsent(lockKey, key -> new CoarseLockState(lockKey));
 
-            return state.acquire(txId, lockKey, lockMode);
+            return state.acquire(txId, lockMode);
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/HeapLockManager.java:177]: It is just a mistake. I am sure the dedicated PR is not necessary.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 2 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:133]: Here, we have to replace an internal timestamp with the timestamp from the method parameter. It is impossible to achieve this by incrementing.

Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:133]: > Do we have any benchmark results here?
No, I did not run a benchmark.
We already estimated that the incrementAndGet has a less impact, either on latency or throughput.

Comment (ascherbakoff) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:133]: incrementAndGet definitely works better under contention

**Comment by vldpyatkov** [comment]
> Doesn't such an update use CAS under the hood? According to the code of `AtomicLongFieldUpdater`, it does:

In my opinion, incrementAndGet has better performance because it does not require a hold in an external loop. Instead of it, we just increment the internal value in the atomic variable.
We already compare this in micro-benchmark.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:133]: > Here, we have to replace an internal timestamp with the timestamp from the method parameter. It is impossible to achieve this by incrementing.

I see, my mistake, thank you for the reply!

File: modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java [hunk 6]
```diff
@@ -75,13 +75,13 @@ public void testTick() {
         assertTimestampEquals(100, new HybridTimestamp(100, 2),
                 () -> clock.update(new HybridTimestamp(60, 1000)));
 
-        assertTimestampEquals(200, new HybridTimestamp(200, 0),
+        assertTimestampEquals(200, new HybridTimestamp(100, 3),
```
Comment (ascherbakoff) [inline modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java:78]: Why this changes are done?
Clock behavior should remain the same.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 1 above)
Comment (ascherbakoff) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:128]: What do you mean by "not changed by jumping up".
Clock here is moved up to current physical time.

**Review by ascherbakoff — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 1 above)
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java]: Yes, it happens the same as in the now() invocation, where we do not notify the clock waiter.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java [hunk 7]
```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
      */
     @Override
     public final HybridTimestamp update(HybridTimestamp requestTime) {
-        while (true) {
-            long now = currentTime();
+        long requestTimeLong = requestTime.longValue();
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:114]: We are using it several times below.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java [hunk 8]
```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
...
 
+        while (true) {
             // Read the latest time after accessing UTC time to reduce contention.
-            long oldLatestTime = latestTime;
+            long oldLatestTime = this.latestTime;
 
-            long newLatestTime = max(requestTime.longValue() + 1, max(now, oldLatestTime + 1));
+            if (oldLatestTime >= requestTimeLong) {
+                return hybridTimestamp(LATEST_TIME.incrementAndGet(this));
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:120]: Benchmark added.

File: modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java (hunk 6 above)
Comment (vldpyatkov) [reply modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java:78]: I do not think that the change violates the clock protocol but increases performance.

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ascherbakoff) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java:114]: This is just a getter.

File: modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java (hunk 6 above)
Comment (ascherbakoff) [reply modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java:78]: I don't understand. That is the reason to change test?

Comment (vldpyatkov) [reply modules/core/src/test/java/org/apache/ignite/internal/hlc/HybridClockTest.java:78]: The reason is we can get the ts bigger than that is provided in parameters by incrementing the old value regardless of the physical ts.

**Review by ascherbakoff — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java (hunk 1 above)
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java]: After thinking a bit, It seems there should be no problem here

**Review by rpuch — APPROVED** [review]
